Crash ``'FunctionDef' object has no attribute 'ancestors'``

### Bug description

When parsing the file here:

https://github.com/openedx/XBlock/blob/59c2f28e32548b8cb78b8f7c733bf64689a9a751/xblock/test/test_core.py

pylint crashed with a ``AttributeError`` and with the following stacktrace:
```
Traceback (most recent call last):
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1034, in _check_files
    self._check_file(get_ast, check_astroid_module, file)
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1069, in _check_file
    check_astroid_module(ast_node)
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1203, in check_astroid_module
    retval = self._check_astroid_module(
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/lint/pylinter.py", line 1250, in _check_astroid_module
    walker.walk(node)
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 75, in walk
    self.walk(child)
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 75, in walk
    self.walk(child)
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 75, in walk
    self.walk(child)
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 72, in walk
    callback(astroid)
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/checkers/refactoring/implicit_booleaness_checker.py", line 142, in visit_compare
    self._check_use_implicit_booleaness_not_comparison(node)
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/checkers/refactoring/implicit_booleaness_checker.py", line 167, in _check_use_implicit_booleaness_not_comparison
    mother_classes = self.base_classes_of_node(target_instance)
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/checkers/refactoring/implicit_booleaness_checker.py", line 214, in base_classes_of_node
    return [instance.name] + [x.name for x in instance.ancestors()]
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/astroid/bases.py", line 134, in __getattr__
    return getattr(self._proxied, name)
AttributeError: 'FunctionDef' object has no attribute 'ancestors'
```

### Configuration

_No response_

### Command used

```shell
`pylint xblock` was the command used.

The project uses Tox, so that command was indirectly run by running `tox -e quality` from the top-level dir of the https://github.com/openedx/XBlock repository.
```


### Pylint output

```shell
************* Module xblock.test.test_plugin
xblock/test/test_plugin.py:68:11: C1803: 'list(...) == []' can be simplified to 'not list(...)' as an empty sequence is falsey (use-implicit-booleaness-not-comparison)
Exception on node <Compare l.747 at 0x114d6d7c0> in file '/Users/jeskew/edx/src/XBlock/xblock/test/test_core.py'
Traceback (most recent call last):
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/utils/ast_walker.py", line 72, in walk
    callback(astroid)
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/checkers/refactoring/implicit_booleaness_checker.py", line 142, in visit_compare
    self._check_use_implicit_booleaness_not_comparison(node)
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/checkers/refactoring/implicit_booleaness_checker.py", line 167, in _check_use_implicit_booleaness_not_comparison
    mother_classes = self.base_classes_of_node(target_instance)
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/pylint/checkers/refactoring/implicit_booleaness_checker.py", line 214, in base_classes_of_node
    return [instance.name] + [x.name for x in instance.ancestors()]
  File "/Users/jeskew/edx/src/XBlock/.tox/quality/lib/python3.8/site-packages/astroid/bases.py", line 134, in __getattr__
    return getattr(self._proxied, name)
AttributeError: 'FunctionDef' object has no attribute 'ancestors'
************* Module xblock.test.test_core
xblock/test/test_core.py:1:0: F0001: Fatal error while checking 'xblock/test/test_core.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/Users/jeskew/Library/Caches/pylint/pylint-crash-2022-01-07-16.txt'. (fatal)
```


### Expected behavior

All the repo code should pass pylint checks. In fact, if pylint is pinned back to `pylint<2.12`, all the repo code *does* pass pylint checks.

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.2
Python 3.8.7 (default, Feb  1 2021, 15:53:07)
[Clang 11.0.3 (clang-1103.0.32.62)]
```


### OS / Environment

macOS Catalina, version 10.15.7
Using iTerm2 terminal

### Additional dependencies

```
#
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    make upgrade
#
appdirs==1.4.4
    # via
    #   -r requirements/django.txt
    #   fs
astroid==2.9.2
    # via
    #   -r requirements/test.in
    #   pylint
    #   pylint-celery
attrs==21.4.0
    # via
    #   hypothesis
    #   pytest
boto3==1.20.30
    # via
    #   -r requirements/django.txt
    #   fs-s3fs
botocore==1.23.30
    # via
    #   -r requirements/django.txt
    #   boto3
    #   s3transfer
click==8.0.3
    # via
    #   click-log
    #   code-annotations
    #   edx-lint
click-log==0.3.2
    # via edx-lint
code-annotations==1.2.0
    # via edx-lint
coverage[toml]==6.2
    # via
    #   -r requirements/test.in
    #   pytest-cov
ddt==1.4.4
    # via -r requirements/test.in
diff-cover==4.0.0
    # via
    #   -c requirements/constraints.txt
    #   -r requirements/test.in
distlib==0.3.4
    # via virtualenv
    # via
    #   -c https://raw.githubusercontent.com/edx/edx-lint/master/edx_lint/files/common_constraints.txt
    #   -r requirements/django.txt
    #   django-pyfs
django-pyfs==3.1.0
    # via -r requirements/django.txt
edx-lint==5.2.1
    # via -r requirements/test.in
filelock==3.4.2
    # via
    #   tox
    #   virtualenv
fs==2.4.14
    # via
    #   -r requirements/django.txt
    #   django-pyfs
    #   fs-s3fs
fs-s3fs==1.1.1
    # via
    #   -r requirements/django.txt
    #   django-pyfs
hypothesis==6.34.2
    # via -r requirements/test.in
inflect==5.3.0
    # via jinja2-pluralize
iniconfig==1.1.1
    # via pytest
isort==5.10.1
    # via pylint
jinja2==3.0.3
    # via
    #   code-annotations
    #   diff-cover
    #   jinja2-pluralize
jinja2-pluralize==0.3.0
    # via diff-cover
jmespath==0.10.0
    # via
    #   -r requirements/django.txt
    #   boto3
    #   botocore
lazy==1.4
    # via -r requirements/django.txt
lazy-object-proxy==1.7.1
    # via astroid
lxml==4.7.1
    # via -r requirements/django.txt
markupsafe==2.0.1
    # via
    #   -r requirements/django.txt
    #   jinja2
mccabe==0.6.1
    # via pylint
mock==4.0.3
    # via -r requirements/test.in
packaging==21.3
    # via
    #   pytest
    #   tox
path==16.2.0
    # via -r requirements/test.in
pbr==5.8.0
    # via stevedore
platformdirs==2.4.1
    # via
    #   pylint
    #   virtualenv
pluggy==1.0.0
    # via
    #   diff-cover
    #   pytest
    #   tox
py==1.11.0
    # via
    #   pytest
    #   tox
pycodestyle==2.8.0
    # via -r requirements/test.in
pygments==2.11.2
    # via diff-cover
pylint==2.12.2
    # via
    #   -r requirements/test.in
    #   edx-lint
    #   pylint-celery
    #   pylint-django
    #   pylint-plugin-utils
pylint-celery==0.3
    # via edx-lint
pylint-django==2.5.0
    # via edx-lint
pylint-plugin-utils==0.7
    # via
    #   pylint-celery
    #   pylint-django
pyparsing==3.0.6
    # via packaging
pytest==6.2.5
    # via
    #   -r requirements/test.in
    #   pytest-cov
    #   pytest-django
pytest-cov==3.0.0
    # via -r requirements/test.in
pytest-django==4.5.2
    # via -r requirements/test.in
python-dateutil==2.8.2
    # via
    #   -r requirements/django.txt
    #   botocore
python-slugify==5.0.2
    # via code-annotations
pytz==2021.3
    # via
    #   -r requirements/django.txt
    #   django
    #   fs
pyyaml==6.0
    # via
    #   -r requirements/django.txt
    #   code-annotations
s3transfer==0.5.0
    # via
    #   -r requirements/django.txt
    #   boto3
simplejson==3.17.6
    # via -r requirements/django.txt
six==1.16.0
    # via
    #   -r requirements/django.txt
    #   edx-lint
    #   fs
    #   fs-s3fs
    #   python-dateutil
    #   tox
    #   virtualenv
sortedcontainers==2.4.0
    # via hypothesis
sqlparse==0.4.2
    # via
    #   -r requirements/django.txt
    #   django
stevedore==3.5.0
    # via code-annotations
text-unidecode==1.3
    # via python-slugify
toml==0.10.2
    # via
    #   pylint
    #   pytest
    #   tox
tomli==2.0.0
    # via coverage
tox==3.24.5
    # via -r requirements/test.in
typing-extensions==4.0.1
    # via
    #   astroid
    #   pylint
urllib3==1.26.7
    # via
    #   -r requirements/django.txt
    #   botocore
virtualenv==20.13.0
    # via tox
web-fragments==2.0.0
    # via -r requirements/django.txt
webob==1.8.7
    # via -r requirements/django.txt
wrapt==1.13.3
    # via astroid

# The following packages are considered to be unsafe in a requirements file:
# setuptools
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -37,6 +37,11 @@ Release date: TBA
 
   Closes #5483
 
+* Fixed crash on properties and inherited class methods when comparing them for
+  equality against an empty dict.
+
+  Closes #5646
+
 * Fixed a false positive for ``assigning-non-slot`` when the slotted class
   defined ``__setattr__``.
 

--- a/doc/whatsnew/2.13.rst
+++ b/doc/whatsnew/2.13.rst
@@ -40,6 +40,11 @@ Other Changes
 * When run in parallel mode ``pylint`` now pickles the data passed to subprocesses with
   the ``dill`` package. The ``dill`` package has therefore been added as a dependency.
 
+* Fixed crash on properties and inherited class methods when comparing them for
+  equality against an empty dict.
+
+  Closes #5646
+
 * By default, pylint does no longer take files starting with ``.#`` into account. Those are
   considered `emacs file locks`_. This behavior can be reverted by redefining the
   ``ignore-patterns`` option.

--- a/pylint/checkers/refactoring/implicit_booleaness_checker.py
+++ b/pylint/checkers/refactoring/implicit_booleaness_checker.py
@@ -107,7 +107,7 @@ class ImplicitBooleanessChecker(checkers.BaseChecker):
         except astroid.InferenceError:
             # Probably undefined-variable, abort check
             return
-        mother_classes = self.base_names_of_base_node(instance)
+        mother_classes = self.base_names_of_instance(instance)
         affected_by_pep8 = any(
             t in mother_classes for t in ("str", "tuple", "list", "set")
         )
@@ -165,7 +165,7 @@ class ImplicitBooleanessChecker(checkers.BaseChecker):
                 target_instance = utils.safe_infer(target_node)
                 if target_instance is None:
                     continue
-                mother_classes = self.base_names_of_base_node(target_instance)
+                mother_classes = self.base_names_of_instance(target_instance)
                 is_base_comprehension_type = any(
                     t in mother_classes for t in ("tuple", "list", "dict", "set")
                 )
@@ -209,17 +209,13 @@ class ImplicitBooleanessChecker(checkers.BaseChecker):
                     )
 
     @staticmethod
-    def base_names_of_base_node(
-        node: Union[bases.UnboundMethod, bases.Instance]
+    def base_names_of_instance(
+        node: Union[bases.Uninferable, bases.Instance]
     ) -> List[str]:
         """Return all names inherited by a class instance or those returned by a function.
 
         The inherited names include 'object'.
         """
-        if isinstance(node, bases.UnboundMethod):
-            return [i.name for i in node.infer_call_result(node)]
         if isinstance(node, bases.Instance):
             return [node.name] + [x.name for x in node.ancestors()]
-
-        # For example for astroid.nodes.Uninferable
         return []

--- a/pylint/checkers/refactoring/implicit_booleaness_checker.py
+++ b/pylint/checkers/refactoring/implicit_booleaness_checker.py
@@ -107,7 +107,7 @@ class ImplicitBooleanessChecker(checkers.BaseChecker):
         except astroid.InferenceError:
             # Probably undefined-variable, abort check
             return
-        mother_classes = self.base_classes_and_return_names(instance)
+        mother_classes = self.base_names_of_base_node(instance)
         affected_by_pep8 = any(
             t in mother_classes for t in ("str", "tuple", "list", "set")
         )
@@ -165,7 +165,7 @@ class ImplicitBooleanessChecker(checkers.BaseChecker):
                 target_instance = utils.safe_infer(target_node)
                 if target_instance is None:
                     continue
-                mother_classes = self.base_classes_and_return_names(target_instance)
+                mother_classes = self.base_names_of_base_node(target_instance)
                 is_base_comprehension_type = any(
                     t in mother_classes for t in ("tuple", "list", "dict", "set")
                 )
@@ -209,7 +209,7 @@ class ImplicitBooleanessChecker(checkers.BaseChecker):
                     )
 
     @staticmethod
-    def base_classes_and_return_names(
+    def base_names_of_base_node(
         node: Union[bases.UnboundMethod, bases.Instance]
     ) -> List[str]:
         """Return all names inherited by a class instance or those returned by a function.

--- a/pylint/checkers/refactoring/implicit_booleaness_checker.py
+++ b/pylint/checkers/refactoring/implicit_booleaness_checker.py
@@ -1,9 +1,9 @@
 # Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 # For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
-from typing import List
+from typing import List, Union
 
 import astroid
-from astroid import nodes
+from astroid import bases, nodes
 
 from pylint import checkers, interfaces
 from pylint.checkers import utils
@@ -107,7 +107,7 @@ class ImplicitBooleanessChecker(checkers.BaseChecker):
         except astroid.InferenceError:
             # Probably undefined-variable, abort check
             return
-        mother_classes = self.base_classes_of_node(instance)
+        mother_classes = self.base_classes_and_return_names(instance)
         affected_by_pep8 = any(
             t in mother_classes for t in ("str", "tuple", "list", "set")
         )
@@ -165,7 +165,7 @@ class ImplicitBooleanessChecker(checkers.BaseChecker):
                 target_instance = utils.safe_infer(target_node)
                 if target_instance is None:
                     continue
-                mother_classes = self.base_classes_of_node(target_instance)
+                mother_classes = self.base_classes_and_return_names(target_instance)
                 is_base_comprehension_type = any(
                     t in mother_classes for t in ("tuple", "list", "dict", "set")
                 )
@@ -209,9 +209,17 @@ class ImplicitBooleanessChecker(checkers.BaseChecker):
                     )
 
     @staticmethod
-    def base_classes_of_node(instance: nodes.ClassDef) -> List[str]:
-        """Return all the classes names that a ClassDef inherit from including 'object'."""
-        try:
-            return [instance.name] + [x.name for x in instance.ancestors()]
-        except TypeError:
-            return [instance.name]
+    def base_classes_and_return_names(
+        node: Union[bases.UnboundMethod, bases.Instance]
+    ) -> List[str]:
+        """Return all names inherited by a class instance or those returned by a function.
+
+        The inherited names include 'object'.
+        """
+        if isinstance(node, bases.UnboundMethod):
+            return [i.name for i in node.infer_call_result(node)]
+        if isinstance(node, bases.Instance):
+            return [node.name] + [x.name for x in node.ancestors()]
+
+        # For example for astroid.nodes.Uninferable
+        return []

--- a/tests/checkers/unittest_refactoring.py
+++ b/tests/checkers/unittest_refactoring.py
@@ -5,10 +5,8 @@ import os
 import signal
 from contextlib import contextmanager
 
-import astroid
 import pytest
 
-from pylint.checkers.refactoring import ImplicitBooleanessChecker
 from pylint.lint import Run
 from pylint.reporters.text import TextReporter
 
@@ -26,52 +24,6 @@ def timeout(timeout_s: float):
     yield
     signal.setitimer(signal.ITIMER_REAL, 0)
     signal.signal(signal.SIGALRM, signal.SIG_DFL)
-
-
-def test_class_tree_detection() -> None:
-    module = astroid.parse(
-        """
-class ClassWithBool(list):
-    def __bool__(self):
-        return True
-
-class ClassWithoutBool(dict):
-    pass
-
-class ChildClassWithBool(ClassWithBool):
-    pass
-
-class ChildClassWithoutBool(ClassWithoutBool):
-    pass
-"""
-    )
-    with_bool, without_bool, child_with_bool, child_without_bool = module.body
-    assert ImplicitBooleanessChecker().base_classes_and_return_names(with_bool) == [
-        "ClassWithBool",
-        "list",
-        "object",
-    ]
-    assert ImplicitBooleanessChecker().base_classes_and_return_names(without_bool) == [
-        "ClassWithoutBool",
-        "dict",
-        "object",
-    ]
-    assert ImplicitBooleanessChecker().base_classes_and_return_names(
-        child_with_bool
-    ) == [
-        "ChildClassWithBool",
-        "ClassWithBool",
-        "list",
-        "object",
-    ]
-    assert ImplicitBooleanessChecker().base_classes_and_return_names(
-        child_without_bool
-    ) == [
-        "ChildClassWithoutBool",
-        "ClassWithoutBool",
-        "dict",
-        "object",
-    ]
 
 
 @pytest.mark.skipif(not hasattr(signal, "setitimer"), reason="Assumes POSIX signals")

--- a/tests/checkers/unittest_refactoring.py
+++ b/tests/checkers/unittest_refactoring.py
@@ -46,23 +46,27 @@ class ChildClassWithoutBool(ClassWithoutBool):
 """
     )
     with_bool, without_bool, child_with_bool, child_without_bool = module.body
-    assert ImplicitBooleanessChecker().base_classes_of_node(with_bool) == [
+    assert ImplicitBooleanessChecker().base_classes_and_return_names(with_bool) == [
         "ClassWithBool",
         "list",
         "object",
     ]
-    assert ImplicitBooleanessChecker().base_classes_of_node(without_bool) == [
+    assert ImplicitBooleanessChecker().base_classes_and_return_names(without_bool) == [
         "ClassWithoutBool",
         "dict",
         "object",
     ]
-    assert ImplicitBooleanessChecker().base_classes_of_node(child_with_bool) == [
+    assert ImplicitBooleanessChecker().base_classes_and_return_names(
+        child_with_bool
+    ) == [
         "ChildClassWithBool",
         "ClassWithBool",
         "list",
         "object",
     ]
-    assert ImplicitBooleanessChecker().base_classes_of_node(child_without_bool) == [
+    assert ImplicitBooleanessChecker().base_classes_and_return_names(
+        child_without_bool
+    ) == [
         "ChildClassWithoutBool",
         "ClassWithoutBool",
         "dict",

--- a/tests/functional/u/use/use_implicit_booleaness_not_comparison.py
+++ b/tests/functional/u/use/use_implicit_booleaness_not_comparison.py
@@ -194,3 +194,46 @@ if data or not data:
 long_test = {}
 if long_test == {        }: # [use-implicit-booleaness-not-comparison]
     pass
+
+
+# Check for properties and uninferable class methods
+# See https://github.com/PyCQA/pylint/issues/5646
+from xyz import AnotherClassWithProperty
+
+
+class ParentWithProperty:
+
+    @classmethod
+    @property
+    def parent_function(cls):
+        return {}
+
+
+class MyClassWithProxy(ParentWithProperty):
+
+    attribute = True
+
+    @property
+    @classmethod
+    def my_property(cls):
+        return {}
+
+    @property
+    @classmethod
+    def my_difficult_property(cls):
+        if cls.attribute:
+            return {}
+        return MyClassWithProxy()
+
+
+
+def test_func():
+    """Some assertions against empty dicts."""
+    my_class = MyClassWithProxy()
+    assert my_class.parent_function == {}  # [use-implicit-booleaness-not-comparison]
+    assert my_class.my_property == {}  # [use-implicit-booleaness-not-comparison]
+
+    # If the return value is not always implicit boolean, don't raise
+    assert my_class.my_difficult_property == {}
+    # Uninferable does not raise
+    assert AnotherClassWithProperty().my_property == {}

--- a/tests/functional/u/use/use_implicit_booleaness_not_comparison.txt
+++ b/tests/functional/u/use/use_implicit_booleaness_not_comparison.txt
@@ -28,3 +28,5 @@ use-implicit-booleaness-not-comparison:160:3:160:20::'numpy_array >= ()' can be 
 use-implicit-booleaness-not-comparison:185:3:185:13::'data == {}' can be simplified to 'not data' as an empty sequence is falsey:UNDEFINED
 use-implicit-booleaness-not-comparison:187:3:187:13::'data != {}' can be simplified to 'data' as an empty sequence is falsey:UNDEFINED
 use-implicit-booleaness-not-comparison:195:3:195:26::'long_test == {}' can be simplified to 'not long_test' as an empty sequence is falsey:UNDEFINED
+use-implicit-booleaness-not-comparison:233:11:233:41:test_func:'my_class.parent_function == {}' can be simplified to 'not my_class.parent_function' as an empty sequence is falsey:UNDEFINED
+use-implicit-booleaness-not-comparison:234:11:234:37:test_func:'my_class.my_property == {}' can be simplified to 'not my_class.my_property' as an empty sequence is falsey:UNDEFINED

